### PR TITLE
gecko_ia2 getLabelElement: cause relationTargetsOfType not to fail

### LINKED ISSUE #8585: pr #8352 causes visible aria-labelledby to be reported twice in Firefox
### Steps to reproduce:
Go to the following address in Firefox:
```
data:text/html,<label id="label">hello</lable><input type="text" aria-labelledby="label" />
```

### Actual behavior:
When reading that line with NVDA, it says "hello hello edit". 

### Expected behavior:
It should suppress reading the label on the edit field itself as the label is already visible before it.
I.e. "hello edit".

### Technical info

PR #8352 introduced this issue.
 
nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp, getLabelElement: the call to IAccessible2_2::get_relationTargetsOfType succeeds, but nTargets is always 0.
Strangely, if we add a log call, or in deed simply reset a wostringstream E.g.
```
wstringstream s;
s.str()=l"";
```
Anywhere in that function (including after the call) The bug goes away.

@jcsteh and I have not managed to find any questionable code in Firefox so far.
But:
* the bug occurs on 64 bit versions of Firefox, 61 stable and hire.
* The bug occurs both on windows 7 and Windows 10.
* the bug goes away if multiprocess in Firefox is turned off, and therefore calls to IAccessible2_2 methods are direct in the content process.
* the bug never occurs on 32 bit Firefox.
 
Compiling with  /analyze /W4 on gecko_ia2.cpp does not report any questionable code except for one small thing:
* a level 4 warning about extra 4 bytes of padding added to the struct in std::unique_ptr that holds IUnknown*, deleter.
But as this is a level 4 warning, I'd be worried if this was in deed the root cause.

However, refactoring the code slightly so we get rid of the deleter and instead use a unique_ptr array of CComPtr<IUnknown>, the bug also goes away.

#### NVDA Installed/portable/running from source:
Installed.

#### NVDA version:
NVDA alpha-15758,2e237f77
#### Windows version:
Windows 10 insider build 17728

#### Name and version of other software in use when reproducing the issue:
Firefox stable 61.0.1 64 bit.

#### Does the issue still occur after restarting your PC?
yes.

#### Have you tried any other versions of NVDA?
yes.

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #8585 

### Summary of the issue:
pr #8352 causes visible aria-labelledby to be reported twice in Firefox.
nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp, getLabelElement: the call to IAccessible2_2::get_relationTargetsOfType succeeds, but nTargets is always 0.

### Description of how this pull request fixes the issue:
Refactor getLabelElement to use a unique_ptr array of CComPtr<IUnknown> objects.
 This time with out a custom deleter. Rather we attach the handed back IUnknowns to the smart pointers and free Gecko's memory block with CoTaskMemFree as soon as possible.

### Testing performed:
Test in issue #8585 

### Known issues with pull request:
None.

### Change log entry:
None (introduced after 2018.2).

Section: New features, Changes, Bug fixes

### CODE REVIEW

**Comment by michaelDCurran** [comment]
This pr also addresses an assertion where boost::optional was used incorrectly. To set its value after initialization, you do not need to dereference the object.

**Comment by jcsteh** [comment]
I've only read the PR comments so far. I assume you don't quite really understand why this fixes it? I mean sure, you're calling CoTaskMemFree earlier, but that still shouldn't have any impact on nTargets; COM has already well and truly returned by that point and shouldn't be able to affect anything. Or is there something I'm missing?

**Comment by michaelDCurran** [comment]
Nope. I don't know any more than that.
But I was able to at least find out that the problem does exist on 
Windows 7 with 64 bit Firefox.
And not at all with 32 bit firefox.

**Review by jcsteh — APPROVED** [review]

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 1]
```diff
@@ -31,33 +31,26 @@ This license can be found at:
...
+	IUnknown** ppUnk=nullptr;
+	long nTargets=0;
+	constexpr int numRelations=2;
+	HRESULT res=element->get_relationTargetsOfType(IA2_RELATION_LABELLED_BY,numRelations,&ppUnk,&nTargets);
+	if(res!=S_OK) return nullptr;
+	// Grab all the returned IUnknowns and store them as smart pointers within a smart pointer array 
+	// so that any further returns will correctly release all the objects. 
+	auto ppUnk_smart=make_unique<CComPtr<IUnknown>[]>(nTargets);
+	for(int i=0;i<nTargets;++i) {
+		ppUnk_smart[i].Attach(ppUnk[i]);
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: If this throws, `CoTaskMemFree` will not be called.

**Pull request merged**